desc/integrals/_laplace.py

```diff
@@ -0,0 +1,401 @@
+        if evl_grid.equiv(Phi_grid):
+            evl_data = Phi_data
+        elif not self._same_grid_phi_src and evl_grid.equiv(src_grid):
+            evl_data = src_data
+        else:
+            evl_data = surface.compute(position, grid=evl_grid)
+
+        errorif(B0 is None and B0n is None)
+        src_data["Bn"] = (
+            B0n
+            if B0n is not None
+            else B0.compute_Bnormal(
+                surface,
+                eval_grid=src_grid,
+                source_grid=src_grid,
+                vc_source_grid=src_grid,
+                chunk_size=chunk_size,
+            )[0]
+        )
+
+        self._B0 = B0
+        self._evl_grid = evl_grid
+        self._interior = interior
+        self._data = {"evl": evl_data, "Phi": Phi_data, "src": src_data}
+        self._transform = {"Phi": Phi_transform, "src": src_transform}
+        self._interpolator = {
+            "evl": get_interpolator(evl_grid, src_grid, src_data, use_dft, **kwargs),
+            "Phi": get_interpolator(Phi_grid, src_grid, src_data, use_dft, **kwargs),
+        }
+
+    def compute_Phi_mn(self, chunk_size=None):
+        """Compute Fourier coefficients of vacuum potential Φ on ∂D.
+
+        Parameters
+        ----------
+        chunk_size : int or None
+            Size to split computation into chunks.
+            If no chunking should be done or the chunk size is the full input
+            then supply ``None``. Default is ``None``.
+
+        Returns
+        -------
+        data : dict
+             Fourier coefficients of Φ on ∂D stored in ``data["Phi"]["Phi_mn"]``.
+
+        """
+        return _compute_Phi_mn(self, chunk_size=chunk_size)
+
+    def _compute_virtual_current(self):
+        """𝐊_vc = -𝐧 × ∇Φ.
+
+        This is the vacuum portion of the virtual casing current.
+        """
+        if "K_vc" in self._data["src"]:
+            return self._data
+
+        Phi_mn = self._data["Phi"]["Phi_mn"]
+        data = self._data["src"]
+        data["Phi_t"] = self._transform["src"].transform(Phi_mn, dt=1)
+        data["Phi_z"] = self._transform["src"].transform(Phi_mn, dz=1)
+        data["K^theta"] = data["Phi_z"] / data["|e_theta x e_zeta|"]
+        data["K^zeta"] = -data["Phi_t"] / data["|e_theta x e_zeta|"]
+        data["K_vc"] = (
+            data["K^theta"][:, jnp.newaxis] * data["e_theta"]
+            + data["K^zeta"][:, jnp.newaxis] * data["e_zeta"]
+        )
+        return self._data
+
+    def compute_vacuum_field(self, chunk_size=None):
+        """Compute magnetic field due to vacuum potential Φ.
+
+        Parameters
+        ----------
+        chunk_size : int or None
+            Size to split computation into chunks.
+            If no chunking should be done or the chunk size is the full input
+            then supply ``None``. Default is ``None``.
+
+        Returns
+        -------
+        data : dict
+             Vacuum field ∇Φ stored in ``data["evl"]["grad(Phi)"]``.
```

Comment:
if you prefer some other api let me know. i was taking care to make these functions pure (recall that is a technical term) so that they play nice with jax

Reply:
ideally each method would be a registered compute function